contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,274 @@
+deploy them and restart the plugins when they fail.
+
+Following this, it makes sense that the user would deploy Device Plugins
+through DaemonSets.
+
+For users installing Kubernetes without using an integrated solution
+such as `kubeadm` they could use the examples that we would put in
+`https://github.com/kubernetes/kubernetes/tree/master/examples/device-plugin`
+
+In the case of integrated solutions such as `kubeadm` we can have them support
+auto-deploying of vetted Device Plugins.
+
+### API Changes
+#### ResourceType
+
+When discovering the devices, Kubelet will be in charge of advertising those
+resources to the API server.
+
+We will advertise each device returned by the Device Plugin in a new structure
+called the ResourceType.
+It is defined as follows:
+
+```golang
+type ResourceType struct {
+	Kind string
+	Name string
+	Quantity resource.Quantity
+	Properties map[string]string
+}
+```
+
+Because the current API (Capacity) can not be extended to support ResourceType,
+we will need to create two new attributes in the NodeStatus structure:
+  * `CapacityV2`: Describing the capacity of the node
+  * `Available`: Describing the available resources
+
+```golang
+type NodeStatus struct {
+	Capacity ResourceList
+	CapacityV2 []ResourceType
+
+	Available[]ResourceType
+}
+```
+
+We also introduce the `Allocated` field in the pod's status so that user
+can know what devices were assigned to the pod. It could also be useful in
+the case of monitoring
+
+```golang
+type PodStatus struct {
+	Allocated[]ResourceType
+}
+```
+
+### Device Plugin
+
+The plugins will have to register themselves with the kubelet
+when they start running and [communicate the API version against
+which they were build](#versioning).
+
+Kubelet will then start interacting with the plugin through the `List`,
+`Watch`, `Allocate` and `Deallocate` functions.
+
+The device plugin will have to mount `/var/lib/kubelet/plugins/device-plugin/kubelet.sock`
+which is the socket that the plugin can use to register itself.
+
+Registration is a simple process where the device plugin communicates it's IP
+address (obtained though the downwards API) and the port on which the gRPC
+server is listening.
+
+When receiving a pod which requests GPUs kubelet will be in charge of:
+  * deciding which device to assign to the pod's containers
+  * advertising the changes to the node's `Available` list
+  * advertising the changes to the pods's `Allocated` list
+  * Calling the `Allocate` function with the list of devices
+
+The scheduler will still be in charge of filtering the nodes which cannot
+satisfy the resource requests.
+
+The typical process will follow the following pattern:
+1. A user submits a pod spec requesting X devices
+2. The scheduler filters the nodes which do not match the resource requests
+3. The pod lands on the node
+
+4. Kubelet decides which device should be assigned to the pod
+5. Kubelet removes the devices from the `Available` pool
```

Comment:
IMO, maintaining `Available` pool can be skipped at kubelet. Wondering how it would be if kubelet just interface between device plugin and scheduler(actually deciding the device after translating resource class)